### PR TITLE
DEV: Load client site settings YML into JS tests

### PR DESCRIPTION
**Pull request by martin-brennan**
Our method of loading a subset of client settings into tests via tests/helpers/site-settings.js can be improved upon. Currently we have a hardcoded subset of the client settings, which may get out of date and not have the correct defaults. As well as this plugins do not get their settings into the tests, so whenever you need a setting from a plugin, even if it has a default, you have to do `needs.setting({ ... })` which is inconvenient.

This commit introduces an ember CLI build step to take the `site_settings.yml` and all the plugin `settings.yml` files, pull out the client settings, and dump them into a variable in a single JS file we can load in our tests, so we have the correct selection of settings and default values in our JS tests. It also fixes many, many tests that were operating under incorrect assumptions or old settings.

**For reviewers:**

Important changes are in ember-cli-build.js, site-settings-plugin.js, helpers/site-settings.js

### CODE REVIEW

**Comment by martin-brennan** [comment]
Tests changed here have been changed to reflect the actual default settings.

File: app/assets/javascripts/discourse/lib/site-settings-plugin.js [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+    let parsed = {};
+
+    this.inputPaths.forEach((path) => {
+      let inputFile;
+      if (path.includes("plugins")) {
+        inputFile = "settings.yml";
+      } else {
+        inputFile = "site_settings.yml";
+      }
+      let file = path + "/" + inputFile;
```
Comment (AndrewPrigorshnev) [inline app/assets/javascripts/discourse/lib/site-settings-plugin.js:34]:
```suggestion
      const file = path + "/" + inputFile;
```

File: app/assets/javascripts/discourse/lib/site-settings-plugin.js [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+      let yaml;
+      try {
+        yaml = fs.readFileSync(file, { encoding: "UTF-8" });
+      } catch (err) {
+        // the plugin does not have a config file, go to the next file
+        return;
+      }
+      let loaded = Yaml.load(yaml, { json: true });
```
Comment (AndrewPrigorshnev) [inline app/assets/javascripts/discourse/lib/site-settings-plugin.js:42]:
```suggestion
      const loaded = Yaml.load(yaml, { json: true });
```

File: app/assets/javascripts/discourse/lib/site-settings-plugin.js [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+    let clientSettings = {};
+    // eslint-disable-next-line no-unused-vars
+    for (const [category, settings] of Object.entries(parsed)) {
+      for (const [setting, details] of Object.entries(settings)) {
+        if (details.client) {
+          clientSettings[setting] = details.default;
+        }
+      }
+    }
+    let contents = `var CLIENT_SITE_SETTINGS_WITH_DEFAULTS  = ${JSON.stringify(
```
Comment (AndrewPrigorshnev) [inline app/assets/javascripts/discourse/lib/site-settings-plugin.js:55]:
```suggestion
    const contents = `var CLIENT_SITE_SETTINGS_WITH_DEFAULTS  = ${JSON.stringify(
```

File: app/assets/javascripts/discourse/lib/site-settings-plugin.js [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+};
+
+module.exports.parsePluginClientSettings = function (discourseRoot) {
+  let settings = [discourseRoot + "/config"];
+
+  if (shouldLoadPluginTestJs()) {
+    settings = settings.concat(glob.sync(discourseRoot + "/plugins/*/config"));
+  }
+
+  let loadedSettings = new SiteSettingsPlugin(settings, "site_settings.yml");
```
Comment (AndrewPrigorshnev) [inline app/assets/javascripts/discourse/lib/site-settings-plugin.js:75]:
```suggestion
  const loadedSettings = new SiteSettingsPlugin(settings, "site_settings.yml");
```

**Review by AndrewPrigorshnev — APPROVED** [review]
Loading default settings into tests is a great improvement!

We may use `const` instead of `let` in some places. I don't think we have a rule about it, but sticking with `const` instead of `let` is a better default, I think.

**Comment by martin-brennan** [comment]
Thanks @AndrewPrigorshnev , yes I agree about the `const`, will clean up that file. I copy/pasted a lot of it from another plugin we have and didn't notice.

File: app/assets/javascripts/discourse/lib/site-settings-plugin.js [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+
+module.exports = function siteSettingsPlugin(...params) {
+  return new SiteSettingsPlugin(...params);
+};
+
+module.exports.parsePluginClientSettings = function (discourseRoot) {
+  let settings = [discourseRoot + "/config"];
+
+  if (shouldLoadPluginTestJs()) {
+    settings = settings.concat(glob.sync(discourseRoot + "/plugins/*/config"));
```
Comment (davidtaylorhq) [inline app/assets/javascripts/discourse/lib/site-settings-plugin.js:72]: For consistency with other plugin-enumerating stuff, we should only include plugin directories which have a `plugin.rb`, and we should exclude plugins which have directory names starting with a `.`.

Easiest thing is probably to lean on the [`discourse-plugins` addon](https://github.com/discourse/discourse/blob/91b991320c267aad6ba88dde26f9fe5729508689/app/assets/javascripts/discourse-plugins/index.js#L86-L94) like this:

https://github.com/discourse/discourse/blob/91b991320c267aad6ba88dde26f9fe5729508689/app/assets/javascripts/discourse/lib/bootstrap-json/index.js#L384-L399

(we'll need to pass a reference to `app` into this function from `ember-cli-build.js`)

**Comment by martin-brennan** [comment]
@davidtaylorhq I made the suggested `pluginInfos` change, can you please just check that again?

**Comment by martin-brennan** [comment]
Single failing test is just a flaky.

**Review by davidtaylorhq — APPROVED** [review]

**Pull request merged**